server/jobs/project_command_output_handler.go

```diff
@@ -68,29 +70,71 @@ type ProjectCommandOutputHandler interface {
 
 	// Cleans up resources for a pull
 	CleanUp(pullInfo PullInfo)
+
+	// Persists job to storage backend and marks operation complete
+	CloseJob(jobID string)
 }
 
 func NewAsyncProjectCommandOutputHandler(
 	projectCmdOutput chan *ProjectCmdOutputLine,
 	logger logging.SimpleLogging,
+	storageBackend StorageBackend,
 ) ProjectCommandOutputHandler {
 	return &AsyncProjectCommandOutputHandler{
 		projectCmdOutput:     projectCmdOutput,
 		logger:               logger,
 		receiverBuffers:      map[string]map[chan string]bool{},
 		projectOutputBuffers: map[string]OutputBuffer{},
 		pullToJobMapping:     sync.Map{},
+		storageBackend:       storageBackend,
+	}
+}
+
+func (p *AsyncProjectCommandOutputHandler) CloseJob(jobID string) {
+	// Mark operation complete
+	p.projectOutputBuffersLock.RLock()
+	outputBuffer := p.projectOutputBuffers[jobID]
+	outputBuffer.OperationComplete = true
+	p.projectOutputBuffers[jobID] = outputBuffer
+	p.projectOutputBuffersLock.RUnlock()
+
+	p.closeActiveChannels(jobID)
+
+	// Persist logs to storage backend
+	ok, err := p.storageBackend.Write(jobID, outputBuffer.Buffer)
```

Comment:
Here, we persist the job synchronously. We could spin up new goroutines for this but I thought it would be good to see how long it actually takes and how much it impacts our response time(after we have our instrumented implementation) before we make premature optimizations!

Reply:
Yeah I like this approach better! I've made the changes accordingly.